src/BreakpointViewer.cpp

```diff
@@ -0,0 +1,621 @@
+	// }
+	assert(index != -1);
+
+	const QString id = breakpoints->getAttribute(index, "id").toString();
+	const QString cmdStr = breakpoints->createRemoveCommand(id);
+
+	qDebug() << "cmdStr =" << cmdStr;
+	auto command = new Command(cmdStr,
+		[this, type, row] (const QString& result) {
+			qDebug() << "result =" << result;
+			createBreakpoint(type, row);
+		}
+	);
+	CommClient::instance().sendCommand(command);
+}
+
+// TODO: move the createRemoveCommand to a session manager
+void BreakpointViewer::removeBreakpoint(Type type, int row)
+{
+	int index = items.indexOf({type, row});
+	// qDebug() << "removeBreakpoint, type" << type << "row" << row;
+	// for (auto item: items) {
+	// 	qDebug() << "item" << item.type << item.row;
+	// }
+	// assert(index != -1);
+
+	QString id = breakpoints->getAttribute(index, "id").toString();
+	QString cmdStr = Breakpoints::createRemoveCommand(id);
+
+	auto command = new Command(cmdStr,
+		[this] (const QString result) {
+			emit contentsUpdated(false);
+	});
+	CommClient::instance().sendCommand(command);
+}
+
+bool BreakpointViewer::processSlotField(int index, const QString& field, qint8& ps, qint8& ss)
+{
+	QStringList s = field.simplified().split("/", Qt::SplitBehaviorFlags::SkipEmptyParts);
+
+	if (s[0].compare("X", Qt::CaseInsensitive)) {
+		bool ok;
+		int value = s[0].toInt(&ok);
+		if (!ok || (value < 0 && value > 3)) {
+			// restore value
+			if (index != -1)
+				ps = breakpoints->getAttribute(index, "ps").toInt();
+			else
+				return false;
+		} else {
+			ps = value;
+		}
+	} else {
+		ps = -1;
+	}
+	if (s.length() == 2 && s[1].compare("X", Qt::CaseInsensitive)) {
+		bool ok;
+		int value = s[1].toInt(&ok);
+		if (!ok || (value < 0 && value > 3)) {
+			if (index != -1)
+				ss = breakpoints->getAttribute(index, "ss").toInt();
+			else
+				return false;
+		} else {
+			ss = value;
+		}
+	} else {
+		ss = -1;
+	}
+	return true;
+}
+
+bool BreakpointViewer::processSegmentField(int index, const QString& field, qint16& segment)
+{
+	QString s = field.simplified();
+
+	if (s.compare("X", Qt::CaseInsensitive)) {
+		bool ok;
+		int value = field.toInt(&ok);
+		if (!ok || (value < 0 && value > 255)) {
+			if (index != -1)
+				segment = breakpoints->getAttribute(index, "segment").toInt();
+			else
+				return false;
+		} else {
+			segment = value;
+		}
+	} else {
+		segment = -1;
+	}
+	return true;
+}
+
+bool BreakpointViewer::processLocationField(int index, BreakpointViewer::Type type,
+		const QString& field, int& begin, int& end)
+{
+	if (type == BreakpointViewer::BREAKPOINT) {
+		int value;
+		if ((value = stringToValue(field.simplified())) != -1) {
+			begin = 0xFFFF & value;
+			end = -1;
+			return true;
+		} else if (index != -1) {
+			begin = breakpoints->getAddress(index);
+			end = -1;
+			return true;
+		}
+		return false;
+	}
+	QStringList s = field.simplified().split(":", Qt::SplitBehaviorFlags::SkipEmptyParts);
+	int value;
+	if (!s.length() || (value = stringToValue(s[0])) == -1) {
```

Comment:
Detail, feel free to ignore:
* `!s.length()`  -> `s.empty()`
Probably doesn't make a difference here, but it more directly expresses the intend.
* `s.length()`  -> `s.size()`
I might be totally wrong, but I think `length()` and `count()` are the old(?) Qt-style names. Modern(?) Qt also has the STL-style name `size()`. At least this makes it more familiar for someone like me, who's more used to STL than to Qt. And "standard" names make it easier to migrate the code in the future.